File: leetcode986/leetcode986_1.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& A, vector<vector<int>>& B) {
        int counter1 = 0, counter2 = 0;
        vector<vector<int>> ans;
        
        while (counter1 < A.size() && counter2 < B.size()) {
            int lo = max(A[counter1][0], B[counter2][0]);
            int hi = min(A[counter1][1], B[counter2][1]);
            
            if (lo <= hi) 
                ans.push_back({lo, hi});
            
            if (A[counter1][1] < B[counter2][1])
                ++counter1;
            else
                ++counter2;
        }
        
        return ans;
    }
};


// Runtime: 48 ms, faster than 91.05% of C++ online submissions for Interval List Intersections.
// Memory Usage: 15.8 MB, less than 80.00% of C++ online submissions for Interval List Intersections.
// time complexity: O(m + n)
// space complexity: O(1)

File: leetcode986/leetcode986_1.java
class Solution {
    public int[][] intervalIntersection(int[][] A, int[][] B) {
      List<int[]> ans = new ArrayList();
      int i = 0, j = 0;
  
      while (i < A.length && j < B.length) {
        // Let's check if A[i] intersects B[j].
        // lo - the startpoint of the intersection
        // hi - the endpoint of the intersection
        int lo = Math.max(A[i][0], B[j][0]);
        int hi = Math.min(A[i][1], B[j][1]);
        if (lo <= hi)
          ans.add(new int[]{lo, hi});
  
        // Remove the interval with the smallest endpoint
        if (A[i][1] < B[j][1])
          i++;
        else
          j++;
      }
  
      return ans.toArray(new int[ans.size()][]);
    }
  }

// Runtime: 2 ms, faster than 99.75% of Java online submissions for Interval List Intersections.
// Memory Usage: 40.2 MB, less than 35.56% of Java online submissions for Interval List Intersections.
// time complexity: O(n + m), which n is length of A and m is length of B
// space complexity: O(n + m)
